scipy/integrate/_quadrature.py

```diff
@@ -822,6 +824,294 @@ def simpson(y, *, x=None, dx=1.0, axis=-1, even=_NoValue):
+    where the widths of the 2 subintervals are:
+    :math:`h_{1} = x_{2} - x_{1}` and :math:`h_{2} = x_{3} - x_{2}`
+
+    Assuming a quadratic relationship over the 3 points, the following 
+    integrals for the subintervals are derived (see [2]_ for derivation 
+    method):
+
+    .. math::
+        \int_{x_{1}}^{x_{2}} y(x) dx\ = \int_{0}^{h_{1}} y(x) dx\
+            = \frac{h_{1}(2h_{1}+3h_{2})}{6(h_{1}+h_{2})} y_{1}
+            + \frac{h_{1}(h_{1}+3h_{2})}{6h_{2}} y_{2}
+            - \frac{h_{1}^{3}}{6h_{2}(h_{1}+h_{2})} y_{3}
+
+    .. math::
+        \int_{x_{2}}^{x_{3}} y(x) dx\ = \int_{h_{1}}^{h_{1}+h_{2}} y(x) dx\
+            = - \frac{h_{2}^{3}}{6h_{1}(h_{1}+h_{2})} y_{1}
+            + \frac{h_{2}(3h_{1}+h_{2})}{6h_{1}} y_{2}
+            + \frac{h_{2}(3h_{1}+2h_{2})}{6(h_{1}+h_{2})} y_{3}
+
+    .. math::
+        \int_{x_{1}}^{x_{3}} y(x) dx\ = \int_{0}^{h_{1}+h_{2}} y(x) dx\
+            = \frac{(h_{1}+h_{2})(2h_{1}-h_{2})}{6h_{1}} y_{1}
+            + \frac{(h_{1}+h_{2})^{3}}{6h_{1}h_{2}} y_{2}
+            - \frac{(h_{1}-2h_{2})(h_{1}+h_{2})}{6h_{2}} y_{3}
+
+    Note that the integral formulae in [1]_ and [2]_ use different variables.
+
+    For samples that are equally spaced the result is exact if the function
+    is a polynomial of order 3 or less [1]_. If the samples are not equally 
+    spaced, then the result is exact only if the function is a polynomial of 
+    order 2 or less.
+
+    References
+    ----------
+    .. [1] Wikipedia page: https://en.wikipedia.org/wiki/Simpson's_rule
+    .. [2] Cartwright, Kenneth V. Simpson's Rule Cumulative Integration with
+            MS Excel and Irregularly-spaced Data. Journal of Mathematical
+            Sciences and Mathematics Education. 12 (2): 1-9
+
+    Examples
+    --------
+    >>> from scipy import integrate
+    >>> import numpy as np
+    >>> import matplotlib.pyplot as plt
+    >>> x = np.linspace(-2, 2, num=20)
+    >>> y = x**2
+    >>> y_int = integrate.cumulative_simpson(y, x=x, initial=0)
+    >>> fig, ax = plt.subplots()
+    >>> ax.plot(x, y_int, 'ro', x, x**3/3 - (x[0])**3/3, 'b-')
+    >>> ax.grid()
+    >>> plt.show()
+
+    """
+    y = np.asarray(y)
+
+    # validate y and axis
+    if axis < -1 or axis >= y.ndim:
+        raise ValueError("If given, `axis` must exist in the shape of `y`.")
+
+    if y.shape[axis] < 3:
+        raise ValueError(
+            "At least 3 points are required along the axis of integration "
+            "to use the composite Simpson's method."
+        )
+
+    if x is not None:
+        x = np.asarray(x)
+        if x.ndim == 1 and y.shape[axis] == x.shape[0]:
+            x_shape = [1] * y.ndim
+            x_shape[axis] = -1
+            x = x.reshape(tuple(x_shape))
+        elif x.shape != y.shape:
+            raise ValueError(
+                "If given, shape of `x` must be the same as `y` or 1-D with "
+                "the same length as `y` along `axis`."
+            )
```

Comment:
Please use [`numpy.broadcast_arrays`](https://numpy.org/doc/stable/reference/generated/numpy.broadcast_arrays.html) and accept any arrays that are mutually broadcastable. Do this near the top of the function. It takes care of converting array-like input into arrays, too.

*Update: Never mind... I guess `simpson` and `cumulative_trapezoid` don't follow normal broadcasting rules. I will make a different suggestion later.*